diff --git a/library/CMakeLists.txt b/library/CMakeLists.txt
index 6aeb385..75b78e1 100644
--- a/library/CMakeLists.txt
+++ b/library/CMakeLists.txt
@@ -157,5 +157,5 @@ endif(USE_SHARED_MBEDTLS_LIBRARY)
 
 add_custom_target(lib DEPENDS mbedcrypto mbedx509 mbedtls)
 if(USE_STATIC_MBEDTLS_LIBRARY AND USE_SHARED_MBEDTLS_LIBRARY)
-    add_dependencies(lib mbedcrypto_static mbedx509_static mbedtls_static)
+    add_dependencies(mbedcrypto mbedcrypto_static mbedx509_static mbedtls_static)
 endif()
