diff --git a/include/mbedtls/config.h b/include/mbedtls/config.h
index 6fc9c772..5138b414 100644
--- a/include/mbedtls/config.h
+++ b/include/mbedtls/config.h
@@ -53,7 +53,10 @@
  *
  * Comment to disable the use of assembly code.
  */
-#define MBEDTLS_HAVE_ASM
+// Disable assembly to avoid the "inline assembly requires more registers than available" error on Android armeabi
+#ifndef __ARM_ARCH_5TE__
+#    define MBEDTLS_HAVE_ASM
+#endif
 
 /**
  * \def MBEDTLS_HAVE_SSE2
@@ -1401,7 +1404,7 @@
  *
  * This modules adds support for the AES-NI instructions on x86-64
  */
-#define MBEDTLS_AESNI_C
+//#define MBEDTLS_AESNI_C
 
 /**
  * \def MBEDTLS_AES_C
@@ -2002,7 +2005,7 @@
  *
  * This modules adds support for the VIA PadLock on x86.
  */
-#define MBEDTLS_PADLOCK_C
+//#define MBEDTLS_PADLOCK_C
 
 /**
  * \def MBEDTLS_PEM_PARSE_C
@@ -2595,6 +2598,6 @@
 #include MBEDTLS_USER_CONFIG_FILE
 #endif
 
-#include "check_config.h"
+#include "mbedtls/check_config.h"
 
 #endif /* MBEDTLS_CONFIG_H */
